### PR TITLE
Update storage docs for content-addressability

### PR DESCRIPTION
**Pull request by tonistiigi**
Signed-off-by: Tonis Tiigi tonistiigi@gmail.com

Mostly the text already described the new system better, as before we really didn't have a separate concept of a layer. All images really only had one layer that was unique to them and a pointer to the parent. 

Big difference is that layer IDs are not currently visible any more. This means you can't easily preview the files on the disk by container/image ID. We could provide some helper tools for that later. AUFS files are still stored in the same place, only the IDs are different.

Other changes that we may want to communicate better:
- Read-only layers are stored by crypo hash. Read-write layers ares stored by random UUID.
- Layer IDs have "sha256:" prefixes (but as they are not visible it doesn't matter much).
- We don't ever create any new empty layers any more. But old images still have them so it would be wrong to just remove them from `ubuntu:15.04` diagrams.

Let me know if there are more docs that need updating.

@moxiegirl @thaJeztah

### CODE REVIEW

**Comment by thaJeztah** [comment]
and /cc @nigelpoulton :-)

I'll have a look tomorrow to see if we need other changes

**Comment by moxiegirl** [comment]
@tonistiigi Thank you.  I'm a bit swamped at the mo..so I'll be looking closer in a day or two. Nigel's commentary here would be super helpful too.

File: docs/userguide/storagedriver/imagesandcontainers.md [hunk 1]
```diff
@@ -148,15 +142,6 @@ One way to do this is using a Dockerfile and the `docker build` command.
     The 4 image layers below it are the exact same image layers the make up the
     ubuntu:15.04 image as their UUIDs match.
```
Comment (thaJeztah) [inline docs/userguide/storagedriver/imagesandcontainers.md]: this may need some rewriting, because those UUIDs are no longer visible

Comment (tonistiigi) [inline docs/userguide/storagedriver/imagesandcontainers.md]: At least one is always visible. All if the image is self-built.

Comment (nigelpoulton) [inline docs/userguide/storagedriver/imagesandcontainers.md]: Oh no... please don't tell me 17924 actually got merged :-P

Comment (thaJeztah) [inline docs/userguide/storagedriver/imagesandcontainers.md]: @nigelpoulton your worst nightmare?

Comment (nigelpoulton) [inline docs/userguide/storagedriver/imagesandcontainers.md]: I vote that we call a code freeze on the whole thing..... progress is overrated ;-)

Just kidding..... I just need some time look into it and get my head around the changes. Once I've done that I think the entire storage driver document set probably needs reviewing and revalidating. We also need a way to reflect the differences between older engines and those with the new stuff. I'll summarize my thoughts on that to @moxiegirl

Comment (tonistiigi) [inline docs/userguide/storagedriver/imagesandcontainers.md]: @nigelpoulton nothing has changed concerning where the layer data is located on disk. There were almost no changes to graphdrivers(except windows). Only thing that is changed is the addressing model from docker objects. I can provide some tools/scripts for people who want to explore the files on disk. There is also a (vague) plan to have something that visualizes the layer sharing, probably as an external tool at first.

**Comment by thaJeztah** [comment]
ping @moxiegirl do you want to move forward with this PR, or do a more extensive rewrite?

**Comment by nigelpoulton** [comment]
@thaJeztah I think it warrants a proper check through the docs in light of the changes. @moxiegirl tell me if you disagree.

I can do this tomorrow if you want.  I'll go through the changes (PR 17924) and vet the docs for any potential changes and make some suggestions.

LMK.

**Comment by moxiegirl** [comment]
@nigelpoulton good to go. I'll email you.

**Comment by thaJeztah** [comment]
I'll close this PR, thanks for starting the work on this @tonistiigi (and looking forward to your changes, @nigelpoulton !)

**Comment by thaJeztah** [comment]
Oh, let me keep it open for now, then @nigelpoulton can carry

**Comment by nigelpoulton** [comment]
@thaJeztah @tonistiigi @moxiegirl Just wanna check..... I'm going through the docs and wanted to check whether engine 1.10.0-dev has this PR merged and is valid for me testing and updating outputs etc in the docs?

I'm also aware of [this](https://github.com/tonistiigi/docker/commit/fc5ba5e3832f748b34c69255cc0d9be69c2e719) from @tonistiigi

**Comment by thaJeztah** [comment]
@nigelpoulton yes, the PR is merge in master, so you can either build master, or download from https://master.dockerproject.org (the experimental channel probably also includes this change)

**Comment by nigelpoulton** [comment]
@tonistiigi A few quick questions for clarification/my understanding....

1. Is this PR merged in 1.10 and later?
2. Are we formally calling images pulled by engines with this PR merged "flat images"?
3. The PR readme says that when graph metadata is migrated as part of an engine upgrade the old metadata is kept. Where is the new and old metadata? E.g. After an upgrade from 1.9.1 to 1.10.0-dev I only have /var/lib/docker/0.0/../.. 
4. Will engines with this PR merged have the "0.0" directory under /var/lib/docker/ or is this just part of the experimental build?
5. Are image ID's based on a hash of the images config file - e.g. /var/lib/docker/0.0/image/aufs/imagedb/content/sha256/<image-id>)
6. From the PR readme: "Images can securely share underlying layers" < Is the new content addressable model more secure than the old way?

Thanks!

@moxiegirl @thaJeztah

**Comment by tonistiigi** [comment]
@nigelpoulton 


> 1. Is this PR merged in 1.10 and later?

Changes that this PR is describing were merged with #17924

> 2. Are we formally calling images pulled by engines with this PR merged "flat images"?

It's a term we use internally to distinguish images that don't require parent chain. There may be a better way to call them.

> 3. The PR readme says that when graph metadata is migrated as part of an engine upgrade the old metadata is kept. Where is the new and old metadata? E.g. After an upgrade from 1.9.1 to 1.10.0-dev I only have /var/lib/docker/0.0/../.. 

We don't change old files. Old image configs are in `graph/` for example but docker does not use them any more.

> 4. Will engines with this PR merged have the "0.0" directory under /var/lib/docker/ or is this just part of the experimental build?

This is part of a userns feature in #12648 not content addressability. I think  currently it is only for experimental, but userns should land on stable for `v1.10`. The directory scheme could change though. One alternative is smth like `/var/lib/docker/aufs/0.0/layers`.

> 5. Are image ID's based on a hash of the images config file - e.g. /var/lib/docker/0.0/image/aufs/imagedb/content/sha256/<image-id>)

Yes, they are the same thing.

> 6. From the PR readme: "Images can securely share underlying layers" < Is the new content addressable model more secure than the old way?

Yes, it is more secure because it avoids all collisions for different data. The sharing aspect itself is not a security feature though.

**Comment by nigelpoulton** [comment]
Much appreciated @tonistiigi

**Comment by icecrime** [comment]
Ping @moxiegirl?

**Comment by moxiegirl** [comment]
@tonistiigi Thanks for the change and the assist on @nigelpoulton questions. LGTM on this.

**Comment by nigelpoulton** [comment]
@tonistiigi is there an easy way to map image layer IDs as shown with `docker pull` to directories under `/var/lib/docker/../..`

Cheers.

**Comment by tonistiigi** [comment]
@nigelpoulton What you see on pull are the IDs in the distribution manifest. If that ID is `sha256:abc`, then `/image/overlay/distribution/diffid-by-digest/sha256/abc` contains the actual layer ID(the one in image conf, also the ID that is stable). If layer ID is `sha256:def` then `image/overlay/layerdb/sha256/def/cache-id` contains the graphdriver ID.

**Comment by nigelpoulton** [comment]
@tonistiigi thanks!

**Pull request merged**